utils/prof_utils.py

```diff
@@ -0,0 +1,1177 @@
+
+    Parameters:
+    -----------
+    x: float
+        The point to evaluate
+    *params: list
+        A list containing three parameters per gaussian component in the order: Amp, Mean, Width
+
+    Returns:
+    --------
+    J: numpy.matrix
+        The Jacobian matrix
+    """
+    def dda(a, b, c, x):
+        return np.exp( -(((x-b)**2) / (2*c**2)) )
+    def ddb(a, b, c, x):
+        return (a/c**2) * (x - b) * np.exp( -(((x-b)**2) / (2*c**2)) )
+    def ddc(a, b, c, x):
+        return (a/c**3) * (x - b)**2 * np.exp( -(((x-b)**2) / (2*c**2)) )
+    J = []
+    for i in range(0, len(params), 3):
+        a = params[i]
+        b = params[i+1]
+        c = params[i+2]
+        mypars = [a, b, c, x]
+        J.append(dda(*mypars))
+        J.append(ddb(*mypars))
+        J.append(ddc(*mypars))
+    J = np.asmatrix(J)
+    return J
+
+#---------------------------------------------------------------
+def fit_gaussian(profile, max_N=6, min_comp_len=0, plot_name=None, alpha=2):
+    """
+    Fits multiple gaussian components to a pulse profile and finds the best number to use for a fit.
+    Will always fit at least one gaussian per profile component.
+    Profile components are defined by find_components().
+    Each gaussian is defined by the following: y = amp * np.exp( -((x - ctr)/wid)**2)
+
+    Parameters:
+    -----------
+    profile: list
+        A list containing the profile data
+    max_N: int
+        OPTIONAL - The maximum number of gaussain components to attempt to fit. Default: 6
+    chi_threshold: float
+        OPTIONAL - The script will stop trying new fits when the reduced chi-squared is within this amount to unity. Default: 0
+    plot_name: string
+        OPTIONAL - If not none, will make a plot of the best fit with this name. Default: None
+
+    Returns:
+    --------
+    [fit, redchisq, best_bic, popt, pcov]: list
+        fit: list
+            The data containing the multi-component gaussian fit to the input profile
+        redchisq: float
+            The reduced chi-sqaured value of the fit
+        best_bic: float
+            The bayesian information criterion for the fit
+        popt: list
+            A list of floats where each 3 numbers describes a single gaussain and are 'ctr', 'amp' and 'wid' respectively
+        pcov: numpy matrix
+            The covariance matrix generated by the curve_fit function
+    """
+    #chi sqaured evaluation
+    def chsq(observed_values, expected_values, err):
+        test_statistic=0
+        for observed, expected in zip(observed_values, expected_values):
+            test_statistic+=((float(observed)-float(expected))/float(err))**2
+        return test_statistic
+
+    #Take noise mean and normalize the profile
+    _, clipped = sigmaClip(profile, alpha=alpha)
+
+    #Check the clipped profile
+
+
+    y = np.array(profile) - np.nanmean(np.array(clipped))
+    max_y = max(y)
+    len_y = len(y)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [local variable 'len_y' is assigned to but never used](https://app.codacy.com/manual/CIRA-Pulsars-and-Transients-Group/vcstools/pullRequest?prid=4645369)